README.md
<!-- The purpose of this document will be to remind ourselves as well as inform other developers of the workflow and processes of this project. A wiki will be created at a later date that hosts all of the information, so we're going to try to keep this document as lightweight as possible. -->

# MIE-MCServer
MIE-MCServer is intended to be a "one-stop shop" for running a private/public Minecraft server on any Linux system. This project should almost completely replace the need for a third-party service for anyone comfortable using terminal. Interacting with this application should be both as simple and as concise as possible.

## Project Setup and Requirements
Installing, running, and developing this project requires Python 3.9 installed on your system. Setup for this project is as easy as cloning the project and calling `python main.py` from the root directory of this project.

## New Features and Helping Out
This project is intended to be completely open-source and everyone is welcome to submit feature requests and implementations as well as any issues they find amongst it. 

If you intend to help out, please see the sections below on Feature Requirements and Pull Requests. If you've found an issue but don't want to solve it yourself for whatever reason, feel free to open an issue, but please read the Issues section below before you do.

### Feature Requirements
As features are added, please keep the following requirements in mind:
- The feature should have a simple accessor, typically in the form of a command argument
- The feature should fill a purpose or role that would typically otherwise be accomplished via one of the following methods:
    - Opening Minecraft, logging into the world, and issuing commands via Minecraft's chat
    - Using a file browser to manually move and/or edit files
    - Issuing other terminal commands to interact with the server and associated files
- More to come?

### Pull Requests
Big or small, we welcome any pull request that stands to make this project even better. When submitting your PR, please use the supplied [pull request template](https://github.com/mietechnologies/MIE-MCServer/blob/main/.github/PULL_REQUEST_TEMPLATE.md). When a PR is reviewed, we will use the [pull request review template](https://github.com/mietechnologies/MIE-MCServer/blob/main/.github/PULL_REQUEST_REVIEW_TEMPLATE.md), so please make sure that you follow the guidelines contained within the template.

### Issues
Bug reports and other issues are also welcome. If you encounter any issues while developing and/or using this project, please open an issue. When submitting your issue, please use the supplied [issue template](link_to_issue_template).

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

### Changed

### Deprecated

### Removed

### Fixed
- Squashed several bugs related to running the project on Raspberry Pi that were created in part by reworking the configuration object, lack of knowledge, and lack of testing.

### Security

## [v1.4.1] - 2022.11.25

### Added
- The user can now update the modpack installed on the server by issuing the `python3 main.py -u` command.

### Fixed
- Corrected several spelling errors and other mistakes surrounding messaging.

## [1.4.0] - 2022.11.25

### Added
- During installation, the user can now declare that they want to install a modded server using Forge.

## [1.3.0] - 2022.11.24

### Added
- The user can now install a datapack using the '-dp {path}' command where path is the absolute path to a zip or directory. 
  - This functionality handles singular compressed (zip) datapacks, a directory with an uncompressed datapack, and a directory that contains multiple compressed datapacks.
- Automatic installation of pip3 and required pip packages to run project when running the project for the first time.
- The user can now designate custom scripts to be ran at start, stop, or clean via `./scripts/custom-command.sh`.
- Server startup monitor to determine if the server was started successfully.

### Changed
- During configuration, the project now fetches the system's total available RAM to intelligently decide what warnings it should display to the user and suggest an appropriate amount of allocated RAM.

### Fixed
- Fixed an issue that was preventing automatic startup scripts from running.

## [1.2.1] - 2022.2.18

### Fixed
- End trimming wasn't being properly called.
- System username was not being fetched correctly.
- Extraneous maintenace arguments were removed.

## [1.2.0] - 2022.02.11

### Added
- Functionality that allows user to schedule, start, or stop maintenance on their Minecraft server or hosting system.
- The ability to run the user's own custom shell scripts placed in `./scripts/custom-command.sh`.
- The user now has the ability to configure settings to post messages into their Discord server.
  - When the system runs into an unhandled error it will now post a message to a Discord server if configured.
- The user can now open a command-line-like session with '-c' (with no argument) to issue a series of server commands.
- The project now asks for and stores the server address during configuration to use in conjuction with RCON.
- Functionality to automatically sign Minecraft's EULA upon user acceptance.
- The user can now update a setting collection with '-uc'.

### Changed
- The reboot process now issues a statement to the players on the server informing them that the server will be saved and stopped in 30 seconds for the server to restart. After the server is stopped, the system will wait for 60 seconds before rebooting.

### Fixed
- When restarting the server, the project would start as if running on a RaspberryPi and start services specific to that system that were causing system errors and generating an egregious number of emails.

### Security
- Any passwords stored within the config are now encrypted.
  - **It is not recommended that the user update any password manually via the config file.**
- Password input is now hidden during setup.

## [1.1.2] - 2022.01.27

### Fixed
- An issue with the Emailer preventing it from sending emails as expected.
- Paper splits entity, poi, and region data for the end regions into separate directories. Previously, only data in the `/region` directory was being removed during end trimming. Now, data for all subdirectories in the appropriate directory is removed.

## [1.1.1] - 2022.01.27

### Added
- The project now fetches the system username to use for scheduling cron jobs.

## [1.1.0] - 2021.12.27

### Added
- Implemented GitHub templates for pull request and pull request review.
- Transferred functionality for running on Raspberry Pi from the [MinePi](https://github.com/Michaelcraun/MinePi) project. 
  - The system's CPU temperature is monitored at intervals configured by the user. 
  - If the system's temperature reaches a critical level, an email is generated to inform the user and the system is restarted.

## [1.0.0] - 2021.12.12


src/main.py
# ****************************** Developer Notes ******************************
# Goal: This script will run and maintain a Minecraft Server, utilizing crontab
# jobs to handle maintenance, and allowing for custom commands to be run from
# the terminal. 
# 1. Setup the basic argument system
#    - No Args: This will be the initial call. It will start everything,
#      from the ground up. It will generate the config file, setup all the
#      needed cron jobs, and start the Minecraft server.
#    - '-mcv', '--mc-version': This will return the currently used Minecraft
#      Server Version.
#    - '-v', '--version': This will return the version of this software.
#    - '-c', '--command': This will accept a string (which will need rapped in
#      quotations) that will run a command on the server (as long as its running)
#    - '-u', '--update': Checks for a Minecraft Server update, and updates if
#      newer version exists.
#    - '-bu', '--backup': This will backup the Minecraft Server
# *****************************************************************************

import argparse
from typing import List
import subprocess
import os
from time import sleep

import command as cmd
from configuration import config
from minecraft.install import Installer
from minecraft.interactions import install_datapack
from minecraft.version import Versioner, UpdateType
from scripts import reboot
from util import files, path, scripting, shell
from util.backup import Backup
from util.date import Date
from util.backup import Backup
from util.date import Date
from util.maintenance import Maintenance
from util.mielib.custominput import bool_input
from util.cron import CronScheduler
from util.backup import Backup
from util.monitor import Monitor
from util.path import project_path
from util.temp import PiTemp
from util.logger import log
from util.date import Date

VERSION = "1.4.1"

def parse(args):
    configuration = config.File(log)

    mc_version = args.minecraft_version
    version = args.version
    command = args.command
    commands = args.commands
    update = args.update
    backup = args.path
    method = args.generate_config
    clean = args.clean
    stop = args.stop
    restart = args.restart
    maintenance_action = args.maintenance
    debug = args.debug
    update_config = args.update_config
    critical_events = None

    if configuration.is_raspberry_pi():
        critical_events = args.critical_events

    if configuration.is_raspberry_pi():
        critical_events = args.critical_events

    running_log = []

    if debug is not False:
        run_debug()
        return

    # Done
    if mc_version is not False:
        running_log.append('-mcv')
        if configuration.minecraft.build is not None:
            log(f"Minecraft Server: {configuration.minecraft.version_str()}")
        else:
            log("Minecraft Server has not been installed yet.")

    # Done
    if version is not False:
        running_log.append('-v')
        log(f"minePi Version v{VERSION}")

    if command is not None:
        running_log.append(f'-c {command}')
        if command == "":
            cmd.run_terminal(configuration)
        else:
            cmd.run_command(command, configuration)

    if update is not None:
        running_log.append('-u')
        update_server(update, configuration)

    if backup:
        running_log.append(f'-bu {configuration.maintenance.path()}')
        cmd.run_command("say System is backing up Minecraft world.", configuration)
        filename = f'world.{Date.strippedTimestamp()}.zip'
        backup_manager = Backup(configuration, log)
        backup_manager.put(Installer.server_dir, configuration.maintenance.path(), filename)

    if method is not None:
        running_log.append(f'-gc {method}')
        generate_config(method, configuration)

    if clean:
        running_log.append('-k')
        scripting.maintenance(configuration)

    if commands:
        running_log.append('-rc')
        scripting.run_user_commands()

    if stop:
        running_log.append('-q')
        cmd.run_command("say The server is being saved, and then stopped " \
            "in 60 seconds.", configuration)
        sleep(60)
        stop_server(configuration)

    if restart:
        running_log.append('-q')
        cmd.run_command("say Saving and stopping server in 30 seconds for system " \
            "restart.", configuration)
        sleep(30)
        stop_server(configuration)
        sleep(60)
        reboot.run()

    if maintenance_action is not None:
        running_log.append(f'-m {maintenance_action}')
        if maintenance_action == 'schedule':
            Maintenance.schedule()
        elif maintenance_action == 'start':
            Maintenance.start()
        else:
            Maintenance.end()

    if configuration.temperature.exists() and critical_events:
        running_log.append('-ce')
        PiTemp(configuration, log).execute()

    if update_config:
        running_log.append('-uc')
        updateConfig(update_config, configuration)

    running_log = __parse_interaction_methods(args, running_log)

    if not running_log:
        run(configuration)

def __parse_interaction_methods(args, running_log: List[str]) -> List[str]:
    datapack_path = args.install_datapack

    if datapack_path:
        running_log.append(f'-dp {datapack_path}')
        install_datapack(datapack_path)

    return running_log

def maintenance(configuration):
    execute_clean_commands(configuration)
    trim_end_regions()
    execute_custom_shell_script()

def execute_clean_commands(configuration):
    log('Running clean up commands...')
    file = project_path('scripts', 'clean-commands.txt')
    cmd.run_terminal(configuration, files.lines_from_file(file))

def execute_command_list(configuration):
    log('Running custom commands...')
    custom_command_file = project_path('scripts', 'commands.txt')
    cmd.run_terminal(configuration, files.lines_from_file(custom_command_file, deleteFetched=True))

def execute_custom_shell_script():
    log('Running custom shell script...')
    custom_script = project_path('scripts', 'custom-command.sh')
    os.chmod(custom_script, 0o755)
    os.system(custom_script)

def trim_end_regions():
    '''
    Cleans the subdirectories related to the end to serve two purposes:
        1. Eliminates lag related to unused and loaded end chunks
        2. Eliminates the need for players to travel very far to find resources
           related to the end
    '''

    log('Trimming the end!')
    log('To keep specific end regions, update the end-regions.txt file in ' \
        'the project root with the regions you would like to keep.')
    log('To determine what region files to keep, see Xisumavoid\'s video at ' \
        'https://www.youtube.com/watch?v=fGlqDBcgmIc')
    
    # Fetch the end region root directory
    root_dir = os.path.dirname(__file__)
    end_dir = os.path.join(root_dir, 'server/world_the_end/DIM1')

    # Fetch the regions the user would like to keep
    end_region_log = os.path.join(root_dir, 'end-regions.txt')
    regions_to_keep = files.lines_from_file(end_region_log)
    filecount = 0

    # Iterate through all subdirectories of the end region root directory
    # and the files contained within each
    for directory in os.listdir(end_dir):
        end_regions = os.path.join(end_dir, directory)
        if os.path.isdir(end_regions):
            dir_count = 0
            for file in os.listdir(end_regions):
                # If the file is not listed in the regions to keep, delete it
                if file not in regions_to_keep:
                    region = os.path.join(end_regions, file)
                    os.remove(region)
                    dir_count += 1
                    filecount += 1
            if dir_count > 0:
                log(f'Removed {dir_count} from {directory}!')
    log(f'Finished trimming the end! Removed {filecount} region(s)!')

def run(configuration: config.File):
    log('Installing prerequisites...')
    __project_preinstalls()

    log("Checking config.yml...")

    if configuration.exists:
        log("Found config.yml")
        setup_crontab(configuration)
        start_server(configuration)
    else:
        log("Did not find config.yml")
        generate_config("manual", configuration)
        setup_crontab(configuration)

        if not configuration.is_modded():
            # This is presumably the first run so the EULA has not yet been
            # accepted, meaning that starting the server WILL fail
            start_server(configuration)

            eula = project_path('server', 'eula.txt')
            if configuration.minecraft.accept_eula():
                files.update(eula, 'eula=false', 'eula=true')
                log('User has accepted Minecraft\'s EULA!')
                log("Server started!")
            else:
                log('User has declined Minecraft\'s EULA!')
                log('Gefore running the Minecraft server, you MUST accept ' \
                    f'Minecraft\'s EULA by updating the { eula } file!')

        if bool_input('Would you like to restart the system to start the server ' \
            'automatically?'):
            reboot.run()
        else:
            log('Please restart the system to start the server.')

def run_debug():
    '''
    A method for running any debug functionality. For PR verification purposes,
    any methods and functionality called here should be left here for your PR to
    allow your PR reviewer to easily pull down and test any changes you make.
    '''
    # Shut off calling server commands for debugging purposes
    cmd.DEBUG = True

    # DO NOT DELETE EITHER OF THE DEBUGGING LINES
    # These are here to give you and testers clear start and stop lines for debugging.
    print('\n****** DEBUGGING STARTED ******\n')
    # Implement any debug functionality below:

    eula = project_path('server', 'eula.txt')
    files.update(eula, 'eula=false', 'eula=true')

    # DO NOT DELETE THE BELOW LINE
    # Deleting this line WILL cause build errors!!
    print('\n***** DEBUGGING FINISHED ******\n')

def startMonitorsIfNeeded(configuration: config.File):
    prog = project_path(filename='main.py')
    scheduler = CronScheduler()

    # Start general system monitors
    monitors = Monitor()
    monitors.start_server_start_monitor(
        timeout=configuration.maintenance.startup_timeout,
        log=log)

    # Temp if on RasPi
    if configuration.temperature.exists():
        log('Start monitor of CPU temp...')
        critical_events_command = f'python {prog} --critical-events'
        scheduler.createRecurringJob(
            '* * * * *',
            critical_events_command,
            'detect_critical_events'
        )

def stopMonitors():
    CronScheduler().removeJob('detect_critical_events')

def start_server(configuration: config.File):
    """
    Starts the server... I mean, the name of the method kinda says it all, no?
    NOTE: This method will start the modded server if configured OR will start the
    Vanilla server otherwise.
    """

    startMonitorsIfNeeded(configuration)
    server = path.project_path('server')
    bootlog = path.project_path('logs', 'bootlog.txt')

    log("Starting server...")
    if configuration.is_modded():
        shell.run(f'{configuration.modded.run_command()} > {bootlog}', server)
    else:
        Installer(configuration, log).install()
        scripting.start(configuration.minecraft.allocated_ram)

def stop_server(configuration):
    stopMonitors()
    scripting.stop()
    cmd.run_command('stop', configuration)

def __project_preinstalls():
    print('Pre-installing needed dependencies to run this command; your ' \
        'input may be required!')

    requirements = project_path('logs', 'requirements.txt')

    os.system('sudo apt install python3-pip')
    os.system('python3 -m pip install pipreqs')
    os.system(f'python3 -m pip install -r {requirements}')

def setup_crontab(configuration: config.File):
    prog = project_path(filename='main.py')
    scheduler = CronScheduler()

    log("Scheduling crontab job 'maintenance.restart'")
    restart_command = f"python3 {prog} --run-commands --stop --restart"
    scheduler.createRecurringJob(
        configuration.maintenance.complete_shutdown,
        restart_command,
        "maintenance.restart"
    )

    # Backup
    log("Scheduling crontab job 'maintenance.backup'")
    backup_command = f"python3 {prog} --backup"
    scheduler.createRecurringJob(
        configuration.maintenance.backup_schedule(),
        backup_command,
        "maintenance.backup"
    )

    # Maintenance
    log("Scheduling crontab job 'maintenance.scripts'")
    maintenance_command = f"python3 {prog} --clean"
    scheduler.createRecurringJob(
        configuration.maintenance.schedule,
        maintenance_command,
        "maintenance.scripts"
    )
    
    # Updates
    log("Scheduling crontab job 'maintenance.update'")
    update_command = f"python3 {prog} --update"
    scheduler.createRecurringJob(
        configuration.maintenance.update_schedule(),
        update_command,
        "maintenance.update"
    )

    # Reboot
    log("Scheduling crontab job 'reboot'")
    reboot_command = f"python3 {prog}"
    scheduler.createRecurringJob(
        "@reboot",
        reboot_command,
        "reboot"
    )

def generate_config(method: str, configuration: config.File):
    if method.lower() == "auto":
        user_response = bool_input("This will override your current " \
        "config.yml, are you sure you want to do that?", default=False)
        if user_response:
            log("Automatically generating a default config.yml")
            configuration.generate()
            log("config.yml generated!")
        else:
            log("Generate config cancelled")
    elif method.lower() == "manual":
        log("Generating user-interactive config.yml", silently=True)
        built = configuration.build()
        if not built:
            log("Generate config cancelled")
        else:
            log("config.yml generated!")
    else:
        print(f"'{method}' is not a valid input. Please consult the help ['-h'] " \
            " menu to learn more. ")

def updateConfig(collection: str, configuration: config.File):
    if collection.lower() == "minecraft":
        configuration.minecraft.reset()
        configuration.minecraft.build()
    elif collection.lower() == "email":
        configuration.email.reset()
        configuration.email.build()
    elif collection.lower() == "maintenance":
        configuration.maintenance.reset()
        configuration.maintenance.build()
    elif collection.lower() == "messaging":
        configuration.messaging.reset()
        configuration.messaging.build()
    elif collection.lower() == "rcon":
        configuration.rcon.build()
    elif collection.lower() == "server":
        configuration.server.build()
    elif collection.lower() == "temperature":
        configuration.temperature.build()

def update_server(override: bool, configuration: config.File):
    if not configuration.exists:
        log("Cancelling... You haven't setup a config.yml file yet. You can " \
            "generate a config file by running the command 'python3 main.py -" \
            "gc'")
    elif configuration.is_modded():
        new_config = configuration.modded.update_modpack()
        configuration.update_section('Modded', new_config)
    else:
        versioner = Versioner(configuration, log)
        update, version = versioner.has_update()

        if update is not UpdateType.NONE:
            should_update = False
            output = f"You have a {update} update. Version {versioner.version_string(version)} ' \
                'is available"
            log(output)

            if update is UpdateType.MAJOR:
                if override.lower() in ["y", "yes", "true"]:
                    should_update = True
                elif not configuration.maintenance.allows_major_udpates():
                    log("Did not update due to your settings. Please update " \
                        "manually.")
                    should_update = False
            else:
                should_update = True

            if should_update:
                Installer.install(override_settings=should_update)

def installRequirements():
    requirements = project_path(filename='requirements.txt')
    with subprocess.Popen('pip install pipreqs') as process:
        process.communicate()
    with subprocess.Popen('pipreqs') as process:
        process.communicate()
    with subprocess.Popen(f'pip install -r {requirements}') as process:
        process.communicate()

def main():
    parser = argparse.ArgumentParser(description="This program is your " \
        "one-stop shop for running and maintaining a Minecraft Server.")

    parser.add_argument('-mcv', '--minecraft-version', help="The version of " \
        "your Minecraft Server.", dest="minecraft_version",
        action="store_true", required=False)

    parser.add_argument('-v', '--version', help="The version of this software.",
        dest="version", action="store_true", required=False)

    parser.add_argument('-c', '--command', help="If ran by itself, without " \
        "the addition of an argument, this will start a terminal which you " \
        "can enter multiple commands in sequence. If an argument is passed " \
        "it will run that command only. In order to stop the terminal you " \
        "can enter '!exit' and it will safely end the terminal session.", 
        dest="command", nargs="?", const="", type=str, required=False)

    parser.add_argument('-u', '--update', help="Checks to see if there is a " \
        "newer version of Minecraft Server. If there is, it will install the " \
        "latest version.", dest="update", nargs="?", const="", required=False)

    parser.add_argument('-bu', '--backup', help="Backup your Minecraft Server", 
        dest="path", action='store_true', required=False)

    parser.add_argument('-k', '--clean', help='Run clean up scripts to help with '\
        'lag on your Minecraft Server.', dest='clean', action='store_true', 
        required=False)

    parser.add_argument('-r', '--restart', help='This command completely ' \
        'restarts your server hardware.', dest='restart', action='store_true',
        required=False)

    parser.add_argument('-q', '--stop', help='This will stop the Minecraft ' \
        'server.', dest='stop', action='store_true', required=False)

    parser.add_argument('-rc', '--run-commands', help='Run the commands ' \
        'line by line in the commands.txt file.', dest='commands',
        action='store_true', required=False)

    parser.add_argument('-gc', '--generate-config', help="This will generate " \
        "the configuration for this program. It will take one of two inputs: " \
        "'auto' or 'manual'. By picking 'auto' we will handle creating a " \
        "config file for you with some default inputs. If you select 'manual'" \
        " we will ask you a series of questions to build you your config. You" \
        " may manually edit or re-generate your config at any time.", 
        dest="generate_config", nargs="?" ,const="auto", type=str,
        required=False)

    parser.add_argument(
        '-m',
        '--maintenance',
        help='Schedule, start, or end maintenance for your Minecraft server ' \
            'or hosting system.',
        const='schedule',
        type=str,
        nargs='?',
        choices=('schedule', 'start', 'end')
    )

    parser.add_argument('-uc', '--update-config', help="This command enables " \
        "the user to update a configuration collection by passing the " \
        "desired collection's name in as a parameter. (i.e. Email, Minecraft, " \
        "etc. [not case sensitive])", nargs='?', dest="update_config",
        type=str, required=False)

    parser.add_argument('-D', '--debug', help='This will run any processes ' \
        'implemented in the runDebug method of main.py. WARN: This command ' \
        'will ignore any and all other commands!', dest='debug', 
        action='store_true', required=False)

    __add_helper_methods(parser)

    config_file = config.File(log)
    if config_file.is_raspberry_pi():
        parser.add_argument(
            '-ce',
            '--critical-events',
            help='Checks for any critical events that may be occuring on your Raspberry Pi.',
            dest='critical_events',
            action='store_true',
            required=False
        )

    parser.set_defaults(func=parse)
    args = parser.parse_args()
    args.func(args)

def __add_helper_methods(parser: argparse.ArgumentParser):
    parser.add_argument('-dp', '--install-datapack', help='This command ' \
        'installs a datapack (or collection of datapacks contained in one ' \
        'directory) when supplied with a file path.', nargs='?',
        dest='install_datapack', type=str, required=False)

if __name__ == "__main__":
    main()
